### PR TITLE
Implement application endpoints

### LINKED ISSUE #68: Implement Repository API applications Endpoints
GET /applications
GET /applications/{application_name}

relates to #16

### PR DESCRIPTION
**Pull request by maschleg**
closes #68

### CODE REVIEW

File: storage/src/main/java/org/zalando/apidiscovery/storage/api/ApplicationResourceController.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.util.UriComponentsBuilder;
+
+@CrossOrigin
+@RestController
+@RequestMapping("/applications")
+public class ApplicationResourceController {
+
+    @Autowired
+    private ApplicationService applicationService;
```
Comment (fmueller) [inline storage/src/main/java/org/zalando/apidiscovery/storage/api/ApplicationResourceController.java]: Please stick to constructor injection. Field based injection should only be used for tests.

File: storage/src/main/java/org/zalando/apidiscovery/storage/api/ApplicationResourceController.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    @GetMapping
+    public ResponseEntity<ApplicationListDto> getApplications(UriComponentsBuilder builder) {
+        List<ApplicationDto> applicationDtoList = applicationService.getAllApplications();
+        return ResponseEntity.ok(new ApplicationListDto(updateApiDefinitionWithLinks(applicationDtoList, builder)));
+    }
+
+    @GetMapping("/{application_name}")
+    public ResponseEntity<ApplicationDto> getApplication(@PathVariable("application_name") String applicationName, UriComponentsBuilder builder) {
+        Optional<ApplicationDto> applicationDto = applicationService.getApplication(applicationName);
+        return applicationDto.map(dto -> ResponseEntity.ok(updateApiDefinitionWithLinks(dto, builder)))
```
Comment (fmueller) [inline storage/src/main/java/org/zalando/apidiscovery/storage/api/ApplicationResourceController.java]: Minor: I'd improve readability by having `.map` right above `orElse`.

File: storage/src/main/java/org/zalando/apidiscovery/storage/api/ApplicationResourceController.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    private List<ApplicationDto> updateApiDefinitionWithLinks(List<ApplicationDto> applicationDtoList, UriComponentsBuilder builder) {
+        applicationDtoList.forEach(
+            applicationDto -> updateApiDefinitionWithLinks(applicationDto, builder)
+        );
+        return applicationDtoList;
+    }
+
+    private ApplicationDto updateApiDefinitionWithLinks(ApplicationDto applicationDto, UriComponentsBuilder builder) {
+        applicationDto.getDefinitions().
```
Comment (fmueller) [inline storage/src/main/java/org/zalando/apidiscovery/storage/api/ApplicationResourceController.java]: That's inconsistent with our code style: the point at the end also belongs on next line.

File: storage/src/main/java/org/zalando/apidiscovery/storage/api/ApplicationService.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import static java.util.stream.Collectors.toList;
+
+@Service
+public class ApplicationService {
+
+    @Autowired
+    private ApplicationRepository applicationRepository;
```
Comment (fmueller) [inline storage/src/main/java/org/zalando/apidiscovery/storage/api/ApplicationService.java]: As above: constructor injection.

File: storage/src/main/java/org/zalando/apidiscovery/storage/api/ApplicationService.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+    public List<ApplicationDto> getAllApplications() {
+        List<ApplicationEntity> applicationEntityList = applicationRepository.findAll();
+        return applicationEntityList.stream()
+            .map(ApplicationEntityToApplicationDtoConverter::toApplicationDto)
+            .collect(toList());
+    }
+
+    public Optional<ApplicationDto> getApplication(String applicationName) {
+        Optional<ApplicationEntity> applicationEntity = Optional.ofNullable(applicationRepository.findOne(applicationName));
```
Comment (fmueller) [inline storage/src/main/java/org/zalando/apidiscovery/storage/api/ApplicationService.java]: I think it is possible to add Optional type directly to `findOne` method of repository. Maybe you have to overwrite it for this case.

File: storage/src/main/java/org/zalando/apidiscovery/storage/api/ApplicationResourceController.java (hunk 1 above)
Comment (maschleg) [inline storage/src/main/java/org/zalando/apidiscovery/storage/api/ApplicationResourceController.java]: yeah right old habit

**Comment by maschleg** [comment]
@fmueller please review changes

File: storage/src/main/java/org/zalando/apidiscovery/storage/api/ApplicationRepository.java [hunk 6]
```diff
@@ -1,17 +1,25 @@
...
     @Query("SELECT DISTINCT a FROM ApplicationEntity a " +
         "INNER JOIN a.apiDeploymentEntities dep " +
         "WHERE dep.api in (:apiEntities)")
     List<ApplicationEntity> findByApiIds(@Param("apiEntities") List<ApiEntity> apiEntities);
 
+    Optional<ApplicationEntity> findOne(String name);
+
+    void deleteAll();
+
+    List<ApplicationEntity> findAll();
```
Comment (mfellner) [inline storage/src/main/java/org/zalando/apidiscovery/storage/api/ApplicationRepository.java]: We could return a Stream here as well: https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#repositories.query-streaming

Comment (fmueller) [inline storage/src/main/java/org/zalando/apidiscovery/storage/api/ApplicationRepository.java]: Ah that's a cool feature, was not aware of this. 👍

Comment (maschleg) [inline storage/src/main/java/org/zalando/apidiscovery/storage/api/ApplicationRepository.java]: "data store specific methods are used to perform the streaming" what does this mean for postgres. [cursor based result sets](https://jdbc.postgresql.org/documentation/head/query.html)?

File: storage/src/main/java/org/zalando/apidiscovery/storage/api/ApplicationRepository.java [hunk 7]
```diff
@@ -1,17 +1,26 @@
...
 
 
     @Query("SELECT DISTINCT a FROM ApplicationEntity a " +
         "INNER JOIN a.apiDeploymentEntities dep " +
         "WHERE dep.api in (:apiEntities)")
     List<ApplicationEntity> findByApiIds(@Param("apiEntities") List<ApiEntity> apiEntities);
 
+    Optional<ApplicationEntity> findOne(String name);
+
+    void deleteAll();
```
Comment (maxim-tschumak) [inline storage/src/main/java/org/zalando/apidiscovery/storage/api/ApplicationRepository.java]: This is used for tests only. In the new setup (see master branch) we execute a SQL script to clean up the data store. Please don't forget to remove this method after merging the master branch ;)

File: storage/src/main/java/org/zalando/apidiscovery/storage/MvcConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+@Configuration
+public class MvcConfiguration extends WebMvcConfigurerAdapter {
+
+    @Autowired
+    private ObjectMapper objectMapper;
+
+    @Override
+    public void configureMessageConverters(final List<HttpMessageConverter<?>> converters) {
+        converters.addAll(messageConverters().getConverters().stream().collect(Collectors.toList()));
```
Comment (maxim-tschumak) [inline storage/src/main/java/org/zalando/apidiscovery/storage/MvcConfiguration.java]: `converters.addAll(messageConverters().getConverters());` should be enough :)

File: storage/src/test/java/org/zalando/apidiscovery/storage/ApplicationResourceComponentTest.java [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+
+    }
+
+    @Test
+    public void shouldReturnAllApplications() throws Exception {
+        ApplicationEntity applicationEntity1 = createApplication("application1");
+        ApiEntity apiEntity = createApiEntity("api1", "v1");
+        createApiDeployment(apiEntity, applicationEntity1);
+
+        System.out.println(mvc.perform(get("/applications")).andReturn().getResponse().getContentAsString());
```
Comment (maxim-tschumak) [inline storage/src/test/java/org/zalando/apidiscovery/storage/ApplicationResourceComponentTest.java]: Please don't forget to remove.

**Comment by maschleg** [comment]
@maxim-tschumak @tkrop please review changes

**Comment by maxim-tschumak** [comment]
:+1:

**Pull request merged**